java/ql/lib/semmle/code/java/security/regexp/NfaUtils.qll

```diff
@@ -0,0 +1,1322 @@
+    }
+
+    bindingset[char]
+    override predicate matches(string char) { not hasChildThatMatches(cc, char) }
+  }
+
+  /**
+   * Holds if the character class escape `clazz` (\d, \s, or \w) matches `char`.
+   */
+  pragma[noinline]
+  private predicate classEscapeMatches(string clazz, string char) {
+    clazz = "d" and
+    char = "0123456789".charAt(_)
+    or
+    clazz = "s" and
+    char = [" ", "\t", "\r", "\n", 11.toUnicode(), 12.toUnicode()] // 11.toUnicode() = \v, 12.toUnicode() = \f
+    or
+    clazz = "w" and
+    char = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789_".charAt(_)
+  }
+
+  /**
+   * An implementation of `CharacterClass` for \d, \s, and \w.
+   */
+  private class PositiveCharacterClassEscape extends CharacterClass {
+    string charClass;
+    RegExpTerm cc;
+
+    PositiveCharacterClassEscape() {
+      isEscapeClass(cc, charClass) and
+      this = getCanonicalCharClass(cc) and
+      charClass = ["d", "s", "w"]
+    }
+
+    override string getARelevantChar() {
+      charClass = "d" and
+      result = ["0", "9"]
+      or
+      charClass = "s" and
+      result = " "
+      or
+      charClass = "w" and
+      if RegExpFlags::isIgnoreCase(cc.getRootTerm())
+      then result = ["a", "z", "_", "0", "9"]
+      else result = ["a", "Z", "_", "0", "9"]
+    }
+
+    override predicate matches(string char) { classEscapeMatches(charClass, char) }
+
+    override string choose() {
+      charClass = "d" and
+      result = "9"
+      or
+      charClass = "s" and
+      result = " "
+      or
+      charClass = "w" and
+      result = "a"
+    }
+  }
+
+  /**
+   * An implementation of `CharacterClass` for \D, \S, and \W.
+   */
+  private class NegativeCharacterClassEscape extends CharacterClass {
+    string charClass;
+
+    NegativeCharacterClassEscape() {
+      exists(RegExpTerm cc |
+        isEscapeClass(cc, charClass) and
+        this = getCanonicalCharClass(cc) and
+        charClass = ["D", "S", "W"]
+      )
+    }
+
+    override string getARelevantChar() {
+      charClass = "D" and
+      result = ["a", "Z", "!"]
+      or
+      charClass = "S" and
+      result = ["a", "9", "!"]
+      or
+      charClass = "W" and
+      result = [" ", "!"]
+    }
+
+    bindingset[char]
+    override predicate matches(string char) {
+      not classEscapeMatches(charClass.toLowerCase(), char)
+    }
+  }
+
+  /** Gets a representative for all char classes that match the same chars as `c`. */
+  CharacterClass normalize(CharacterClass c) {
+    exists(string normalization |
+      normalization = getMormalizationString(c) and
+      result =
+        min(CharacterClass cc, string raw |
+          getMormalizationString(cc) = normalization and cc = CharClass(raw)
+        |
+          cc order by raw
+        )
+    )
+  }
+
+  /** Gets a string representing all the chars matched by `c` */
+  private string getMormalizationString(CharacterClass c) {
```

Comment:
Should this be `getNormalizationString`?

Reply:
Yes it should!